File: clients/ts/tests/common/harness/meteoraDlmmSwap2.ts
import type { Address } from "@solana/addresses";

import { assertTailLength, venues, type VenueWindow } from "../../../src/index.js";
import { slotAt, type Fixture } from "../fixture.js";
import { key } from "../keys.js";
import { assertProgram } from "./index.js";

const { MAX_BIN_ARRAYS, PROGRAM_ADDRESS, resolve: build } = venues.meteoraDlmmSwap2;

function optional(pubkey: Address): Address | undefined {
  return pubkey === PROGRAM_ADDRESS ? undefined : pubkey;
}

export function resolve(fixture: Fixture): VenueWindow {
  assertProgram(fixture, PROGRAM_ADDRESS);
  // slot 13 memo, slot 14 event authority, slot 15 program-again are module constants.
  const binArrays: ReadonlyArray<Address> = fixture.slots.slice(16).map((slot) => slot.pubkey);
  assertTailLength(binArrays, 1, MAX_BIN_ARRAYS);
  return build(
    {
      pool: slotAt(fixture, 0).pubkey,
      binArrayBitmapExtension: optional(slotAt(fixture, 1).pubkey),
      reserveX: slotAt(fixture, 2).pubkey,
      reserveY: slotAt(fixture, 3).pubkey,
      userTokenIn: slotAt(fixture, 4).pubkey,
      userTokenOut: slotAt(fixture, 5).pubkey,
      tokenXMint: slotAt(fixture, 6).pubkey,
      tokenYMint: slotAt(fixture, 7).pubkey,
      oracle: slotAt(fixture, 8).pubkey,
      hostFeeIn: optional(slotAt(fixture, 9).pubkey),
      user: slotAt(fixture, 10).pubkey,
      tokenXProgram: slotAt(fixture, 11).pubkey,
      tokenYProgram: slotAt(fixture, 12).pubkey,
    },
    binArrays,
  );
}

export function reachableAccountCounts(): Array<number> {
  const counts: Array<number> = [];
  for (let length = 1; length <= MAX_BIN_ARRAYS; length += 1) {
    const binArrays: ReadonlyArray<Address> = Array.from({ length }, (_, index) =>
      key(101 + index),
    );
    assertTailLength(binArrays, 1, MAX_BIN_ARRAYS);
    for (const bitmapExtension of [undefined, key(90)]) {
      for (const hostFeeIn of [undefined, key(91)]) {
        counts.push(
          build(
            {
              pool: key(1),
              binArrayBitmapExtension: bitmapExtension,
              reserveX: key(2),
              reserveY: key(3),
              userTokenIn: key(4),
              userTokenOut: key(5),
              tokenXMint: key(6),
              tokenYMint: key(7),
              oracle: key(8),
              hostFeeIn,
              user: key(9),
              tokenXProgram: key(10),
              tokenYProgram: key(11),
            },
            binArrays,
          ).accountCount,
        );
      }
    }
  }
  return counts;
}

File: clients/rust/turk-router/tests/common/harness/meteora_dlmm_swap2.rs
use solana_pubkey::Pubkey;
use turk_router::venues::meteora_dlmm_swap2::{
    resolve as resolve_window, MeteoraDlmmSwap2Accounts, MAX_BIN_ARRAYS, PROGRAM_ID,
};
use turk_router::venues::PubkeyTail;
use turk_router::VenueWindow;

use crate::common::fixture::Fixture;

fn optional(pubkey: Pubkey) -> Option<Pubkey> {
    if pubkey == PROGRAM_ID {
        None
    } else {
        Some(pubkey)
    }
}

pub fn resolve(fixture: &Fixture) -> VenueWindow {
    assert_eq!(fixture.program_id, PROGRAM_ID);

    let accounts = MeteoraDlmmSwap2Accounts {
        pool: fixture.slot(0).pubkey,
        bin_array_bitmap_extension: optional(fixture.slot(1).pubkey),
        reserve_x: fixture.slot(2).pubkey,
        reserve_y: fixture.slot(3).pubkey,
        user_token_in: fixture.slot(4).pubkey,
        user_token_out: fixture.slot(5).pubkey,
        token_x_mint: fixture.slot(6).pubkey,
        token_y_mint: fixture.slot(7).pubkey,
        oracle: fixture.slot(8).pubkey,
        host_fee_in: optional(fixture.slot(9).pubkey),
        user: fixture.slot(10).pubkey,
        token_x_program: fixture.slot(11).pubkey,
        token_y_program: fixture.slot(12).pubkey,
    };
    // slot(13) memo, slot(14) event_authority, slot(15) program-again are module constants.
    let bin_array_keys: Vec<Pubkey> = fixture.slots[16..].iter().map(|slot| slot.pubkey).collect();
    let bin_arrays = PubkeyTail::<MAX_BIN_ARRAYS>::try_from_slice(&bin_array_keys)
        .expect("fixture bin tail length");

    resolve_window(accounts, bin_arrays)
}

pub fn reachable_account_counts() -> Vec<u8> {
    fn key(byte: u8) -> Pubkey {
        Pubkey::new_from_array([byte; 32])
    }

    let bin_pool: [Pubkey; MAX_BIN_ARRAYS] = [
        key(101),
        key(102),
        key(103),
        key(104),
        key(105),
        key(106),
        key(107),
        key(108),
    ];

    let mut counts = Vec::new();
    for len in 1..=MAX_BIN_ARRAYS {
        let bin_arrays =
            PubkeyTail::<MAX_BIN_ARRAYS>::try_from_slice(&bin_pool[..len]).expect("len in range");
        for bitmap_extension in [None, Some(key(90))] {
            for host_fee_in in [None, Some(key(91))] {
                let accounts = MeteoraDlmmSwap2Accounts {
                    pool: key(1),
                    bin_array_bitmap_extension: bitmap_extension,
                    reserve_x: key(2),
                    reserve_y: key(3),
                    user_token_in: key(4),
                    user_token_out: key(5),
                    token_x_mint: key(6),
                    token_y_mint: key(7),
                    oracle: key(8),
                    host_fee_in,
                    user: key(9),
                    token_x_program: key(10),
                    token_y_program: key(11),
                };
                counts.push(resolve_window(accounts, bin_arrays).account_count());
            }
        }
    }
    counts
}
